File: 2020/3.py
#!/usr/bin/env python3
import math

with open("3.input") as file:
    data = [row.strip() for row in file.readlines()]

    x = 0
    y = 0
    t = 0

    for y in range(1, len(data)):
        x = (x + 3) % len(data[0])
        if data[y][x] == '#':
            t += 1

    print(t)

    step = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]
    trees = []
    for s in step:
        t = 0
        x = 0
        y = 0
        while y < len(data):
            if data[y][x] == '#':
                t += 1
            x = (x + s[0]) % len(data[0])
            y += s[1]
        trees.append(t)

    print(trees[1])
    print(math.prod(trees))

File: 2023/3.dart
#!/usr/bin/env dart

import 'dart:io';

void scanScematic(List<String> data) {
  Map<String, int> parts = {};
  Map<String, List<int>> gears = {};

  data.asMap().forEach((y, row) {
    for (RegExpMatch match in RegExp(r'\d+').allMatches(row)) {
      for (int sy = y - 1; sy <= y + 1; sy++) {
        for (int sx = match.start - 1; sx <= match.end; sx ++) {
          if (sx < 0 || sy < 0 || sx >= row.length || sy >= data.length) continue;

          if (data[sy].codeUnitAt(sx) != 46 && (data[sy].codeUnitAt(sx) < 48 || data[sy].codeUnitAt(sx) > 57)) {
            parts[[match.start, y].toString()] = int.parse(match.group(0)!);

            if (data[sy][sx] == '*')
              gears.putIfAbsent([sx, sy].toString(), () => []).add(int.parse(match.group(0)!));
          }
        }
      }
    }
  });

  print(parts.values.reduce((value, element) => value + element));
  print(gears.values.where((element) => element.length > 1).map((gear) => gear.reduce((a, b) => a * b)).reduce((a, b) => a +b));
}

void main() {
  List<String> data = new File("3.input")
      .readAsLinesSync()
      .toList();
  scanScematic(data);
}